include/enoki/texture.h

```diff
@@ -337,53 +385,63 @@ template <typename Value, size_t Dimension> class Texture {
-        step[3] = select(pos_f >= -1.f && pos_f < 0.f,                1,  step[3]);
-        step[3] = select(pos_f >= m_shape_opaque - 2 && pos_i <= m_shape_opaque - 2, 1, step[3]);
-        PosF pos_a = pos_f - pos_i;
-        pos_i = clamp(pos_i, 0, PosI(m_shape_opaque) - 1);
 
-        const auto compute_weight = [&pos_a](uint32_t dim) -> Array4 {
-            const Value &alpha = pos_a[dim];
-            Value alpha2 = alpha * alpha,
-                  alpha3 = alpha2 * alpha;
-            Value multiplier = rcp(6.f);
-            return multiplier * Array4(
-                - alpha3 + 3.f * alpha2 - 3.f * alpha + 1.f,
-                3.f * alpha3 - 6.f * alpha2 + 4.f,
-                -3.f * alpha3 + 3.f * alpha2 + 3.f * alpha + 1.f,
-                alpha3
-            );
-        };
+        // `offset[k]` controls the k-th offset for any dimension.
+        // With cubic B-Spline, it is by default [-1, 0, 1, 2].
+        Array4 offset(-1, 0, 1, 2);
 
-        UInt32 index;
+        InterpPosI pos_i_w(0);
         if constexpr (Dimension == 1) {
-            index = pos_i.x();
+            for (uint32_t ix = 0; ix < 4; ix++) {
+                pos_i_w[0][ix] = offset[ix] + pos_i.x();
+            }
         } else if constexpr (Dimension == 2) {
-            index = fmadd(pos_i.y(), m_shape_opaque.x(), pos_i.x());
-            for (uint32_t idx = 0; idx < 4; ++idx)
-                step[idx][1] *= m_shape_opaque.x();
+            for (uint32_t ix = 0; ix < 4; ix++) {
+                for (uint32_t iy = 0; iy < 4; iy++) {
+                    pos_i_w[0][iy * 4 + ix] = offset[iy] + pos_i.x();
+                    pos_i_w[1][ix * 4 + iy] = offset[iy] + pos_i.y();
+                }
+            }
         } else if constexpr (Dimension == 3) {
-            index = fmadd(
-                fmadd(pos_i.z(), m_shape_opaque.y(), pos_i.y()),
-                m_shape_opaque.x(),
-                pos_i.x());
-            for (uint32_t idx = 0; idx < 4; ++idx) {
-                step[idx][1] *= m_shape_opaque.x();
-                step[idx][2] *= m_shape_opaque.x() * m_shape_opaque.y();
+            for (uint32_t ix = 0; ix < 4; ix++) {
+                for (uint32_t iy = 0; iy < 4; iy++) {
+                    for (uint32_t iz = 0; iz < 4; iz++) {
+                        pos_i_w[0][iz * 16 + iy * 4 + ix] =
+                            offset[iz] + pos_i.x();
+                        pos_i_w[1][ix * 16 + iz * 4 + iy] =
+                            offset[iz] + pos_i.y();
+                        pos_i_w[2][iy * 16 + ix * 4 + iz] =
+                            offset[iz] + pos_i.z();
+                    }
+                }
             }
         }
+        pos_i_w = wrap(pos_i_w);
+
+        PosF pos_a = pos_f - pos_i;
+
+        const auto compute_weight = [&pos_a](uint32_t dim) -> Array4 {
+            const Value &alpha = pos_a[dim];
+            Value alpha2 = alpha * alpha, alpha3 = alpha2 * alpha;
```

Comment:
We tend to avoid multiple declaration on a single line.
```suggestion
            Value alpha2 = alpha * alpha, 
                      alpha3 = alpha2 * alpha;
```